### DISCUSSION TITLE
Optimistic persist

### DISCUSSION THREAD

**Discussion by erano067** [pmndrs/zustand#2497] [Ideas]
Hi,
I was thinking of adding an optimistic feature to the persist middleware, in general, what I wish for is that for any async persist the middleware will clone the current state into `oldState`, change the UI, and then make the appropriate call to the async storage(backend/indexedDb, etc..) with a `.catch(error=>...)` that does a rollback to the by `set(oldState)`

In the options we can add an `onError` callback that will be called when reverting back, so users can pop a toast/log it, etc..

What you guys think?

**Comment by alterhuman** [comment]
> Hi,
> 
> I was thinking of adding an optimistic feature to the persist middleware, in general, what I wish for is that for any async persist the middleware will clone the current state into `oldState`, change the UI, and then make the appropriate call to the async storage(backend/indexedDb, etc..) with a `.catch(error=>...)` that does a rollback to the by `set(oldState)`
> 
> 
> 
> In the options we can add an `onError` callback that will be called when reverting back, so users can pop a toast/log it, etc..
> 
> 
> 
> What you guys think?

Did you end up creating one?
